libs/networking/http_helpers/http_helpers.ml

```diff
@@ -49,22 +49,54 @@ let set_client_ref v = if not !in_mock_context then client_ref := Some v
  * compiling/linking a package OCaml requires a choice of a virtual module
  * implementation, and it can't defer to the thing that's using the package.
  *)
+
 module Make (Lwt_platform : sig
   val run : 'a Lwt.t -> 'a
 end) =
 struct
-  let rec get_async ?(headers = []) caps url =
-    Logs.debug (fun m -> m "GET on %s" (Uri.to_string url));
-    (* This checks to make sure a client has been set *)
-    (* Instead of defaulting to a client, as that can cause *)
-    (* Hard to debug build and runtime issues *)
+  (* Respect the proxy!*)
+  (* Cohttp doesn't https://github.com/mirage/ocaml-cohttp/issues/459 *)
+  (* https://github.com/0install/0install/blob/6c0f5c51bc099370a367102e48723a42cd352b3b/ocaml/zeroinstall/http.cohttp.ml#L232-L256 *)
+  let get_proxy uri =
+    let proxy_uri_env =
+      match Uri.scheme uri with
+      | Some "http" -> Some "HTTP_PROXY"
+      | Some "https" -> Some "HTTPS_PROXY"
+      | _ -> None
+    in
+    let proxy_uri = Option.bind proxy_uri_env Sys.getenv_opt in
+    match proxy_uri with
+    | Some proxy_uri -> Some (Uri.of_string proxy_uri)
+    | None -> None
+
+  (* Why this wrapper function? Client.call takes a uri, and some other things *)
+  (* and then makes a Request.t with said uri and sends that request to the same uri *)
+  (* By using Client.callv, we can make a request that has some uri, but then really *)
+  (* send it to a different uri. This is used for proxying requests *)
+  let call_client ?(body = `Empty) ?(headers = []) ?(chunked = false) meth url =
     let module Client =
       (val match !client_ref with
            | Some client -> client
            | None -> failwith "HTTP client not initialized")
     in
     let headers = Header.of_list headers in
-    let%lwt response, orig_body = Client.get ~headers url in
+    let req = Cohttp.Request.make_for_client ~headers ~chunked meth url in
+    let stream_req = Lwt_stream.of_list [ (req, body) ] in
+    (* Send the request to the proxy, not the original url, if it's set *)
+    let url = Option.value (get_proxy url) ~default:url in
+    let%lwt responses_stream = Client.callv url stream_req in
```

Comment:
I don't understand why we're using `callv` instead of just `call` here. Could you explain?

Reply:
added a comment while trying to fix CI tests. Will push in a bit, but tldr call makes a request with uri, then sends said request to the uri. Callv sends a list of requests to a uri, so we can make a request with a uri = server1, then use callv to send it to uri = server2, in order to support proxies

Reply:
Ok. Could we use https://mirage.github.io/ocaml-cohttp/cohttp-lwt-unix/Cohttp_lwt_unix/Connection/index.html instead? `callv` appears to be deprecated.

Reply:
Didn't see it was deprecated, will use that

Reply:
nevermind, looks like it's only deprecated in 6.0, which is still in beta https://github.com/mirage/ocaml-cohttp/blob/v5.3.0/cohttp-lwt-unix/src/cohttp_lwt_unix.ml. 5.3.0 (our version) doesn't have this module yet.

Reply:
Society when versions. Well, could you leave a note in that case?